src/main/java/org/opensearch/plugin/insights/core/service/categorizer/QueryShapeGenerator.java

```diff
@@ -195,11 +225,60 @@ static String buildSortShape(List<SortBuilder<?>> sortBuilderList, Boolean showF
      * @return String: comma separated list with leading space in square brackets
      * Ex: " [my_field, width:5]"
      */
-    static String buildFieldDataString(NamedWriteable builder) {
+    static String buildFieldDataString(NamedWriteable builder, Metadata metadata, Set<String> searchIndices) {
         List<String> fieldDataList = new ArrayList<>();
         if (builder instanceof WithFieldName) {
-            fieldDataList.add(((WithFieldName) builder).fieldName());
+            String fieldName = ((WithFieldName) builder).fieldName();
+            fieldDataList.add(fieldName);
+            String fieldType = getFieldType(fieldName, metadata, searchIndices);
+            if (fieldType != null) {
+                fieldDataList.add(fieldType);
+            }
         }
         return " [" + String.join(", ", fieldDataList) + "]";
     }
+
+    /**
+     * Method to get a field's type
+     * @return String field type
+     */
+    static String getFieldType(String fieldName, Metadata metadata, Set<String> searchIndices) {
+        if (metadata == null) {
+            return null;
+        }
+        for (String searchIndex : searchIndices) {
+            IndexMetadata indexMetadata = metadata.index(searchIndex);
+            if (indexMetadata == null) {
+                continue;
+            }
+            MappingMetadata mappingMetadata = indexMetadata.mapping();
+            if (mappingMetadata == null) {
+                continue;
+            }
+            Map<String, Object> sourceMap = mappingMetadata.getSourceAsMap();
+
+            String fieldType = findFieldType((Map<String, Object>) sourceMap.get("properties"), fieldName);
+            if (fieldType != null) {
+                return fieldType;
+            }
+        }
+        return null;
+    }
```

Comment:
Can we inject the `IndicesServices` into QueryInsightsPlugin and do something like below:

```
MappedFieldType mappedFieldType = indicesService.indexService(index).mapperService().fieldType(fieldName)
if (mappedFieldType != null) {
    return mappedFieldType.typeName();
}
```

Reply:
We might need to update ```results.getSuccessfulResults().map(result -> result.getSearchShardTarget().getIndex()``` to ```results.getSuccessfulResults().map(result -> result.getSearchShardTarget().getShardId().getIndex()``` for getting the `Index` object instead of `String`

Reply:
`QueryInsightsPlugin` is instantiated in core in `node.java` with the [createComponents()](https://github.com/opensearch-project/query-insights/blob/main/src/main/java/org/opensearch/plugin/insights/QueryInsightsPlugin.java#L66) method. This method `createComponents()` overrides the `Plugin/TelemetryAwarePlugin` class. We will need to overload createComponents to add IndicesServices.

Reply:
TelemetryAwarePlugin is already one example of overloading createComponents() to inject telemetry related classes to relevant plugins. We would need something similar.